File: stack/max_boundary_subarrays/max_boundary_subarrays.cpp
#include <vector>
#include <unordered_map>
#include <deque>
#include <stack>
using namespace std;

long long count_subarrays(vector<int> &nums)
{ // LeetCode Q.3113.
    unordered_map<int, deque<int>> nums2indices;

    vector<int> prev_bigger_indices(nums.size(), -1);
    stack<pair<int, int>> decreasing_stack; // Format: {idx, num}.

    long long subarrays_count = 0;
    for (int idx = 0; idx < nums.size(); idx++)
    {
        int num = nums[idx];
        if (nums2indices.find(num) == nums2indices.end())
            nums2indices[num] = {};
        nums2indices[num].push_back(idx);

        while (!decreasing_stack.empty())
        {
            auto [prev_idx, prev_num] = decreasing_stack.top();
            if (prev_num > num)
            {
                prev_bigger_indices[idx] = prev_idx;
                break;
            }
            decreasing_stack.pop();
        }
        decreasing_stack.push({idx, num});

        while (nums2indices[num].front() < prev_bigger_indices[idx])
        {
            nums2indices[num].pop_front();
        }

        subarrays_count += nums2indices[num].size();
    }

    return subarrays_count;
}

File: stack/3113_hard/max_boundary_subarrays.py

def count_subarrays(nums: list[int]) -> int:  # LeetCode Q.3113.
    nums2indices: dict[int, list[int]] = dict()

    prev_bigger_indices: list[int] = [-1] * len(nums)
    decreasing_stack: list[tuple[int, int]] = []  # Format: (idx, num).

    subarrays_count = 0

    for idx, num in enumerate(nums):
        if num not in nums2indices.keys():
            nums2indices.update({num: []})
        nums2indices[num].append(idx)

        while decreasing_stack and decreasing_stack[-1][1] <= num:
            decreasing_stack.pop()

        if decreasing_stack:
            prev_bigger_indices[idx] = decreasing_stack[-1][0]

        decreasing_stack.append((idx, num))

        while nums2indices[num][0] < prev_bigger_indices[idx]:
            nums2indices[num].pop(0)

        subarrays_count += len(nums2indices[num])

    return subarrays_count
